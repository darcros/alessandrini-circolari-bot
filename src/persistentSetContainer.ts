import { readFile, writeFile, mkdir } from "fs/promises";
import { dirname } from "path";

export class PersistentSetContainer<T> {
  private set: Set<T> = null;
  private readonly filePath: string;

  constructor(filePath: string) {
    this.filePath = filePath;
    this.load();
  }

  public async get(): Promise<Set<T>> {
    if (this.set === null) {
      this.set = await this.load();
    }

    return this.set;
  }

  private async load(): Promise<Set<T>> {
    try {
      const buf = await readFile(this.filePath);
      const arr = JSON.parse(buf.toString());
      return new Set(arr);
    } catch (_) {
      return new Set();
    }
  }

  public async save(): Promise<void> {
    const urls = [...this.set.values()];
    const jsonString = JSON.stringify(urls, null, 2);

    await mkdir(dirname(this.filePath), { recursive: true });
    await writeFile(this.filePath, jsonString);
  }
}
