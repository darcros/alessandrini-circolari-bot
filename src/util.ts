export function fixMultinePadding(s: string) {
  return s.split('\n')
    .map(s => s.trim())
    .join('\n');
}

export function requireEnv(name: string) {
  const envVar = process.env[name];

  if (!envVar) {
    const message = `
      VARIBILI D'AMBIENTE MANCANTI!
      Questa variabile d'ambiente non Ã¨ stata settata: ${name}.
      Puoi settarla in due modi:
      1. crea un file chiamato .env nella stessa cartella di package.json e aggiungi questa riga "${name}=valore"
      2. settala tramite il tuo sistema operativo (con bash, CMD, ecc..)
    `;

    throw new Error(fixMultinePadding(message));
  }

  return envVar;
}
