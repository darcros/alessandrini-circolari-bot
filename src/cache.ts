import { PersistentSetContainer } from './persistentSetContainer';
import { getEnv } from './util';

const CACHE_PATH = getEnv('CACHE_PATH', './data/cache.json');

const urlSet = new PersistentSetContainer<string>(CACHE_PATH);

export async function wasSent(url: string): Promise<boolean> {
  const set = await urlSet.get();
  return set.has(url);
}

export async function confirmSent(url: string): Promise<void> {
  const set = await urlSet.get();
  set.add(url);
  await urlSet.save();
}
