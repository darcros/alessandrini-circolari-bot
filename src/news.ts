import axios from 'axios';
import cheerio from 'cheerio';
import { parse } from 'date-fns';
import { URL } from 'url';
import { wasSent } from './cache';

function scrapeNewsList(page: string) {
  const $ = cheerio.load(page);

  const urls: string[] = [];

  $('.view-content > .views-row').each((_i, tag) => {
    const newsElement = $(tag);

    const titleElement = newsElement.find('.views-field-title > .field-content > a');
    const url = titleElement.attr('href');

    urls.push(url);
  });

  return urls;
}

export interface Attachemnt {
  name: string;
  url: string;
}

// TODO: scrape text
interface ScrapedNews {
  title: string;
  id: string;
  date: Date;
  attachments: Attachemnt[];
}

function scrapeNewsPage(page: string): ScrapedNews {
  const $ = cheerio.load(page);

  const title = $('#page-title').text();

  const dateString = $('.field.field-name-field-data-emissione > .field-item > .date-display-single').text();

  // la data passata come ultimo argoemnto serve a prendere i dati che non vengono specificati esplicitamente dalla stringa,
  // ovvero le ore, minuti, secondi, millisecondi
  // usare la data 0 li imposta tutti a 0
  const date = parse(dateString, 'dd/MM/yyyy', new Date(0));

  const id = $('.field.field-name-field-circolare-protocollo > .field-item').text();

  // NOTE: tra .field-item e .file c'è uno spazion e non un >
  // non è fatto a caso, è perchè ci sono altri elementi in mezzo
  const attachments: Attachemnt[] =
    $('.field.field-name-field-circ-all-riservati > .field-item .file > a')
      .toArray()
      .map((tag) => {
        const name = $(tag).text();
        const url = $(tag).attr('href');
        return { name, url };
      });

  return {
    title,
    date,
    id,
    attachments,
  };
}

async function filterNotSent(urls: string[]): Promise<string[]> {
  const promises = urls.map(async (url) => ({
    url,
    sent: await wasSent(url),
  }));

  const objs = await Promise.all(promises);
  return objs
    .filter(({ sent }) => !sent)
    .map(({ url }) => url);
}

export interface News {
  title: string;
  id: string;
  date: Date;
  url: string;
  absoluteUrl: string;
  attachments: Attachemnt[];
}

export async function scrapeNews(newsListPageUrl: string): Promise<News[]> {
  const { data: newsListPage } = await axios.get<string>(newsListPageUrl);

  const urls = scrapeNewsList(newsListPage);
  console.info(`Trovati ${urls.length} link`);

  const newUrls = await filterNotSent(urls);
  console.info(`Trovati ${newUrls.length} nuovi link`);

  const promises: Promise<News>[] = newUrls.map(async (url) => {
    const absoluteUrl = new URL(url, newsListPageUrl).toString();
    const { data: newsPage } = await axios.get<string>(absoluteUrl);
    const scraped = scrapeNewsPage(newsPage);

    return { ...scraped, url, absoluteUrl };
  });

  // Promise.all() fallisce se una promessa qualsiasi fallisce,
  // questo non fallisce mai, al massimo restituice un array vuoto
  const news = (await Promise.allSettled(promises))
    .flatMap(x => {
      if (x.status === 'rejected') {
        console.error('Errore con una circolare', x.reason);
      }

      return x.status === 'fulfilled' ? [x.value] : []
    });
  console.info(`scaricate ${news.length} nuova circolari`);

  // ordina dalla più vecchi alla più nuova
  return news.reverse();
}
