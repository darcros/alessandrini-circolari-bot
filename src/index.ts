import { format } from 'date-fns';
import * as dotenv from 'dotenv';
import { Telegraf } from 'telegraf';
import { confirmSent } from './cache';
import { scrapeNews } from './news';

import { getEnv, fixMultinePadding } from './util';

// Carica variabili d'ambiente da .env
dotenv.config();

const TELEGRAM_TOKEN = getEnv('TELEGRAM_TOKEN');
const TELEGRAM_CHANNEL_ID = getEnv('TELEGRAM_CHANNEL_ID');
const BASE_URL = getEnv('BASE_URL', 'https://www.alessandrinimainardi.edu.it/categoria/circolari');

async function main() {
  console.info('Ottengo elenco circolari...');
  const newsList = await scrapeNews(BASE_URL);
  console.info(`Ottenute ${newsList.length} circolari.`);

  const bot = new Telegraf(TELEGRAM_TOKEN);

  console.info(`Invio ${newsList.length} messaggi`);
  for (const news of newsList) {
    const message = `
      *üìÑ Circolare*
      [${news.title}](${news.absoluteUrl})

      üìí Numero: ${news.id}
      üìÜ Data: ${format(news.date, 'dd/MM/yyyy')}

      ${news.attachments.length > 0 ? 'üîó Allegati' : ''}
      ${news.attachments
        .map((attachment) => `‚ñ™Ô∏è[${attachment.name}](${attachment.url})`)
        .join('\n')}
    `;

    await bot.telegram.sendMessage(
      TELEGRAM_CHANNEL_ID,
      fixMultinePadding(message),
      {
        parse_mode: 'Markdown',
        disable_web_page_preview: true,
      }
    );
    await confirmSent(news.url);
  }
  console.info('finito');
}

main();
