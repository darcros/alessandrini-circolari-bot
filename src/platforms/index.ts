import { Cache } from '../cache';
import { News } from '../news';
import { discordPlatform } from './discord';
import { telegramPlatform } from './telegram';

export interface Bot {
  readonly platformName: string;

  send(news: News[]): Promise<void>;
  send(news: News): Promise<void>;
}

export interface Platform {
  readonly name: string;

  isEnabled: () => boolean;
  initialize: (cache: Cache) => Bot;
}

const platforms = [telegramPlatform, discordPlatform];

export function getEnabledBots(cache: Cache): Bot[] {
  return platforms
    .filter((platform) => {
      const enabled = platform.isEnabled();
      if (!enabled) {
        console.info(`La piattaforma ${platform.name} Ã¨ disabilitata`);
      }

      return enabled;
    })
    .map((platform) => platform.initialize(cache));
}
