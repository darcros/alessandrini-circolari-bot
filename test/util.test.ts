import { getEnv } from '../src/util';

describe('environment variables', () => {
  beforeAll(() => {
    process.env['foo'] = 'something';
    delete process.env['bar'];
  });

  test('gets defined environemnt variable', () => {
    const env = getEnv('foo');
    expect(env).toBeDefined();
  });

  test('gets default values when environemnt variable is undefined', () => {
    const env = getEnv('bar', 'default');
    expect(env).toBe('default');
  });

  test('thows when getting undefined envornment variable without a default', () => {
    expect(() => getEnv('bar')).toThrowErrorMatchingSnapshot();
  });
});
