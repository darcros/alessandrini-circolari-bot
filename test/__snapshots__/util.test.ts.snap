// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`environment variables thows when getting undefined envornment variable without a default 1`] = `
"
VARIBILI D'AMBIENTE MANCANTI!
Questa variabile d'ambiente non Ã¨ stata settata: bar.
Puoi settarla in due modi:
1. crea un file chiamato .env nella stessa cartella di package.json e aggiungi questa riga \\"bar=valore\\"
2. settala tramite il tuo sistema operativo (con bash, CMD, ecc..)
"
`;
